bridge/engine/EngineStreamDirector.h

```diff
@@ -338,82 +338,86 @@ class EngineStreamDirector
 
     inline bool shouldForwardSsrc(const size_t toEndpointIdHash, const uint32_t ssrc)
     {
-        const auto pinMapItr = _pinMap.find(toEndpointIdHash);
-        if (pinMapItr != _pinMap.end())
-        {
-            if (isSsrcFromParticipant(pinMapItr->second, ssrc))
-            {
-                const auto result = isParticipantHighestActiveQuality(pinMapItr->second, toEndpointIdHash, ssrc);
-                DIRECTOR_LOG("shouldForwardSsrc toEndpointIdHash %lu ssrc %u: %c, pin target high quality",
-                    "EngineStreamDirector",
-                    toEndpointIdHash,
-                    ssrc,
-                    result ? 't' : 'f');
-                return result;
-            }
+        size_t pinnedQuality, unpinnedQuality;
+        const auto viewedByParticipantStreamItr = _participantStreams.find(toEndpointIdHash);
 
-            const auto lowQualitySsrcsItr = _lowQualitySsrcs.find(ssrc);
-            const auto result =
-                lowQualitySsrcsItr != _lowQualitySsrcs.end() && lowQualitySsrcsItr->second != toEndpointIdHash;
+        if (viewedByParticipantStreamItr == _participantStreams.end())
+        {
+            return false;
+        }
 
-            DIRECTOR_LOG("shouldForwardSsrc toEndpointIdHash %lu ssrc %u: %c, non pin target low quality",
-                "EngineStreamDirector",
-                toEndpointIdHash,
-                ssrc,
-                result ? 't' : 'f');
+        getVideoQualityLimits(viewedByParticipantStreamItr->second, pinnedQuality, unpinnedQuality);
 
-            return result;
-        }
+        DIRECTOR_LOG("shouldForwardSsrc toEndpointIdHash %lu ssrc %u: max pinned quality: %d, max unpinned quality: %d",
+            "EngineStreamDirector",
+            toEndpointIdHash,
+            ssrc,
+            pinnedQuality,
+            unpinnedQuality);
 
-        const auto lowQualitySsrcsItr = _lowQualitySsrcs.find(ssrc);
-        const auto midQualitySsrcsItr = _midQualitySsrcs.find(ssrc);
-        if (lowQualitySsrcsItr == _lowQualitySsrcs.end() && midQualitySsrcsItr == _midQualitySsrcs.end())
+        if (isSsrcFromParticipant(toEndpointIdHash, ssrc))
```

Comment:
Move this check to before `getVideoQualityLimits`. `pinnedQuality` and `unpinnedQuality` is not useful when this is `true` so we can save the calculation cost of them and also when the participant is alone in a conference (this will be always true) and the log above will print random values for `pinnedQuality` and `unpinnedQuality` (as `getVideoQualityLimits` will return without set any values).

Reply:
Agree.